File: 78-subsets/78-subsets.cpp
class Solution {
public:
    void helper(int i,vector<int> &nums,vector<int> ds,vector<vector<int>> &res){
        if(i==nums.size())
        {
            res.push_back(ds);
            return;
        }
        ds.push_back(nums[i]);
        helper(i+1,nums,ds,res);
        ds.pop_back();
        helper(i+1,nums,ds,res);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> ds;
        helper(0,nums,ds,res);
        return res;
    }
};

File: 78-subsets/78-subsets.java
class Solution {
    
    
    public List<List<Integer>> subsets(int[] nums) {
    List<List<Integer>> res=new ArrayList<>();
    helper(0,nums,new ArrayList<>(),res);
    return res;
    }
    private void helper(int i,int []nums,List<Integer> ds,List<List<Integer>> res)
    {
        if(i==nums.length)
        {
            res.add(new ArrayList<>(ds));
            return;
        }
        ds.add(nums[i]);
        helper(i+1,nums,ds,res);
        ds.remove(ds.size()-1);
        helper(i+1,nums,ds,res);
        
    }
    
}